oqsprov/oqsx.h

```diff
@@ -19,41 +19,97 @@
 #  include <openssl/core.h>
 #  include <openssl/e_os2.h>
 
+/* Extras for OQS extension */
+
+#define ON_ERR_SET_GOTO(condition, ret, code, gt) \
+    if ((condition)) {                            \
+        printf("ON_ERR_CONDITION: %d, setting code: %d\n", condition, code); fflush(stdout);   \
+        (ret) = (code);                           \
+        goto gt;                                  \
+    }
+
+#define ON_ERR_GOTO(condition, gt) \
+    if ((condition)) {                        \
+        printf("ON_ERR_CONDITION: %d\n", condition); fflush(stdout);   \
+        goto gt;                              \
+    }
+
+#define ECP_NAME(secbits, oqsname) \
+    (secbits == 128 ? "p256_" #oqsname "" : \
+     secbits == 192 ? "p384_" #oqsname "" : \
+                      "p521_" #oqsname "")
+#define ECX_NAME(secbits, oqsname) \
+    (secbits == 128 ? "x25519_" #oqsname "" : \
+                        "x448_" #oqsname "")
+
 typedef struct prov_oqs_ctx_st {
     const OSSL_CORE_HANDLE *handle;
     OSSL_LIB_CTX *libctx;         /* For all provider modules */
 //    BIO_METHOD *corebiometh; // for the time being, do without BIO_METHOD
 } PROV_OQS_CTX;
 
-
 PROV_OQS_CTX *oqsx_newprovctx(OSSL_LIB_CTX *libctx, const OSSL_CORE_HANDLE *handle);
 void oqsx_freeprovctx(PROV_OQS_CTX *ctx);
 # define PROV_OQS_LIBCTX_OF(provctx) (((PROV_OQS_CTX *)provctx)->libctx)
 
 #include "oqs/oqs.h"
 
+struct oqsx_kex_info_st {
+    int nid_kex;
+    int nid_kex_crv;
+    int raw_key_support;
+    size_t kex_length_public_key;
+    size_t kex_length_private_key;
+    size_t kex_length_secret;
+};
+
+typedef struct oqsx_kex_info_st OQSX_KEX_INFO;
+
+struct oqsx_evp_ctx_st {
+    EVP_PKEY_CTX *kex;
+    EVP_PKEY *kexParam;
+    const OQSX_KEX_INFO *kex_info;
+};
+
+typedef struct oqsx_evp_ctx_st OQSX_EVP_CTX;
+
 typedef union {
-    OQS_SIG *s;
-    OQS_KEM *k;
-} OQS_KEY;
+    OQS_SIG *sig;
+    OQS_KEM *kem;
+} OQSX_QS_CTX;
+
+struct oqsx_provider_ctx_st {
+    OQSX_QS_CTX oqsx_qs_ctx;
+    OQSX_EVP_CTX *oqsx_evp_ctx;
+};
+
+typedef struct oqsx_provider_ctx_st OQSX_PROVIDER_CTX;
+
+enum oqsx_key_type_en {
+    KEY_TYPE_SIG, KEY_TYPE_KEM, KEY_TYPE_ECP_HYB_KEM, KEY_TYPE_ECX_HYB_KEM
+};
+
+typedef enum oqsx_key_type_en OQSX_KEY_TYPE;
 
 struct oqsx_key_st {
     OSSL_LIB_CTX *libctx;
     char *propq;
-    unsigned int iskem:1;
-    OQS_KEY key;
+    OQSX_KEY_TYPE keytype;
+    OQSX_PROVIDER_CTX oqsx_provider_ctx;
+    size_t numkeys;
     size_t privkeylen;
     size_t pubkeylen;
     char *oqs_name;
     char *tls_name;
     _Atomic int references;
+    void **privkey_ptr;
```

Comment:
This naming may be confusing ("is privkey_ptr just a pointer to privkey? why have both?"): If I get it right, "privkey_ptr" rather is "component_privkey" and "privkey" _is_ the true (possibly hybrid/composite) private key (?). Same with pubkey(s).

Reply:
Yes, correct, the privkey_ptr elements point to locations in the main key. I will rename them as you suggest. The reason for the "privkey_ptr"/"component_privkey" is that the KEM code can access them using their "keyslot", without having to deal with any offsets in the main "privkey"

Reply:
Changed names in https://github.com/open-quantum-safe/oqs-provider/pull/16/commits/9bd8e14a2ea54c1ed2c79380c40a22ade30e78c3